File: app/src/main/java/com/king/superdemo/fragments/AppRightFragment.java
package com.king.superdemo.fragments;

import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.king.recyclerviewlibrary.CommonAdapter;
import com.king.recyclerviewlibrary.CommonItem;
import com.king.recyclerviewlibrary.CommonRecyclerView;
import com.king.superdemo.R;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class AppRightFragment extends BaseFragment{

    private CommonRecyclerView mRecyclerView;
    private CommonAdapter mCommonAdapter;
    private Context mContext;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        mContext = context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public void setArguments(@Nullable Bundle args) {
        super.setArguments(args);
        Bundle bundle = getArguments();
        if (bundle != null) {
            int position = bundle.getInt("position");
            List<CommonItem> itemList = getItemList(position);
            mCommonAdapter.setCommonItemList(itemList);
            mCommonAdapter.notifyDataSetChanged();
        }

    }

    private List<CommonItem> getItemList(int position) {
        switch (position) {
            case 0 : return  getAllApps();
            case 1 : return getSystemApps();
            case 2 : return getLauncherApps();
            default: return null;
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.app_right_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mRecyclerView = view.findViewById(R.id.app_right_fragment_recycler_view);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(view.getContext()));
        mCommonAdapter = new CommonAdapter(view.getContext());
        mCommonAdapter.setCommonItemList(getAllApps());
        mRecyclerView.setAdapter(mCommonAdapter);
    }

    //获取所有应用
    private  List<CommonItem>  getAllApps() {
        PackageManager packageManager = mContext.getPackageManager();
        List<PackageInfo> packageInfoList = packageManager.getInstalledPackages(0);
        List<CommonItem> commonItemList = mCommonAdapter.covertToCommonHolder(packageInfoList);
        return commonItemList;
    }

    //获取系统应用
    private List<CommonItem> getSystemApps() {
        PackageManager packageManager = mContext.getPackageManager();
        List<PackageInfo> packageInfoList = packageManager.getInstalledPackages(0)
                .stream()
                .filter(packageInfo -> (packageInfo.applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) == ApplicationInfo.FLAG_SYSTEM)
                .collect(Collectors.toList());
        List<CommonItem> commonItemList = mCommonAdapter.covertToCommonHolder(packageInfoList);
        return commonItemList;
    }

    //获取桌面上的应用
    private List<CommonItem> getLauncherApps() {
        PackageManager packageManager = mContext.getPackageManager();
        Intent intent = new Intent();
        intent.addCategory(Intent.CATEGORY_LAUNCHER);
        intent.setAction(Intent.ACTION_MAIN);
        List<ResolveInfo> resolveInfoList = packageManager.queryIntentActivities(intent, 0);

        List<CommonItem> commonItemList = mCommonAdapter.covertToCommonHolder(resolveInfoList);
        return commonItemList;
    }
}

File: app/src/main/java/com/king/superdemo/fragments/AppRightFragment.kt
package com.king.superdemo.fragments

import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.king.recyclerviewlibrary.CommonAdapter
import com.king.recyclerviewlibrary.CommonItem
import com.king.recyclerviewlibrary.CommonRecyclerView
import com.king.superdemo.R
import com.king.superdemo.activities.AppManagerActivity
import kotlinx.coroutines.*
import java.util.stream.Collectors

class AppRightFragment : BaseFragment() {
    private lateinit var mRecyclerView: CommonRecyclerView
    private lateinit var mCommonAdapter: CommonAdapter
    private var mContext: Context? = null

    private lateinit var sharedViewModel: AppManagerActivity.AppSharedViewModel
    override fun onAttach(context: Context) {
        super.onAttach(context)
        mContext = context
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sharedViewModel = ViewModelProvider(requireActivity()).get(AppManagerActivity.AppSharedViewModel::class.java)
        sharedViewModel.data.observe(this){
            val position = it.second
            refreshList(position)
        }
    }

    override fun setArguments(args: Bundle?) {
        super.setArguments(args)
        val bundle = arguments
        bundle?.let {
            val position = it.getInt("position")
            refreshList(position)
        }
    }

    private fun refreshList(position: Int){
        //使用协程获取数据,否则再UI线程中获取数据会阻塞UI线程,当数据量较大时,会有明显的卡顿
        GlobalScope.launch {
            Log.i("wq", "onViewCreated: GlobalScope id="+ Thread.currentThread().name);
            val list = when (position) {
                0 -> allApps
                1 -> systemApps
                2 -> thirdApps
                3 -> launcherApps
                else -> null
            }
            //使用withContext切换到主线程更新UI
            withContext(Dispatchers.Main) {
                mCommonAdapter .setCommonItemList(list)
                mCommonAdapter.notifyDataSetChanged()}
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.app_right_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mRecyclerView = view.findViewById(R.id.app_right_fragment_recycler_view)
        mRecyclerView.setLayoutManager(LinearLayoutManager(view.context))
        mCommonAdapter = CommonAdapter(view.context)
        mRecyclerView.adapter = mCommonAdapter
        refreshList(0)
    }

    //获取所有应用
    private val allApps: List<CommonItem>
        private get() {
            val packageManager = mContext!!.packageManager
            val packageInfoList = packageManager.getInstalledPackages(0)
            return mCommonAdapter.covertToCommonHolder(packageInfoList)
        }

    //获取系统应用
    private val systemApps: List<CommonItem>
        private get() {
            val packageManager = mContext!!.packageManager
            val packageInfoList = packageManager.getInstalledPackages(0)
                    .stream()
                    .filter { it.applicationInfo.flags and ApplicationInfo.FLAG_SYSTEM == ApplicationInfo.FLAG_SYSTEM }
                    .collect(Collectors.toList())
            return mCommonAdapter.covertToCommonHolder(packageInfoList)
        }

    //获取第三方应用
    private val thirdApps: List<CommonItem>
        private get() {
            val packageManager = mContext!!.packageManager
            val packageInfoList = packageManager.getInstalledPackages(0)
                    .stream()
                    .filter { it.applicationInfo.flags and ApplicationInfo.FLAG_SYSTEM != ApplicationInfo.FLAG_SYSTEM }
                    .collect(Collectors.toList())
            return mCommonAdapter.covertToCommonHolder(packageInfoList)
        }

    //获取桌面上的应用
    private val launcherApps: List<CommonItem>
        private get() {
            val packageManager = mContext!!.packageManager
            val intent = Intent()
            intent.addCategory(Intent.CATEGORY_LAUNCHER)
            intent.action = Intent.ACTION_MAIN
            val resolveInfoList = packageManager.queryIntentActivities(intent, 0)
            return mCommonAdapter.covertToCommonHolder(resolveInfoList)
        }
}